### ISSUE TITLE
How to configure client endpoint_url?

### ISSUE THREAD

**Issue by jasonlockporcelane** [numberoverzero/bloop#37]
Hi, 
I'm looking to use DynamoDB local with bloop and was wondering what would be the best way to have bloop configure the boto3 client with a new endpoint_url?

Thanks for the help!

**Comment by numberoverzero** [comment]
Thanks for the question!  I want to add support for DynamoDB local (really, any custom boto3 client) but hadn't prioritized it yet.

From [this SO post](http://stackoverflow.com/questions/31948742/localhost-endpoint-to-dynamodb-local-with-boto3) I think it would be easiest if I made `bloop.Client` take a boto3 client, and then it's the usual pattern to construct an engine.

I can get a revision up in a few hours.  How do you feel about this configuration?

```python
import boto3
import bloop

# Talk to DynamoDB Local
boto_client = ddb = boto3.client("dynamodb", endpoint_url="http://localhost:8000")
bloop_client = bloop.Client(boto_client=boto_client)
engine = bloop.Engine(client=bloop_client)
```

- [ ] bloop.Client takes an optional boto_client
- [ ] bloop.Engine takes an optional bloop_client
- [ ] bloop.Engine.bind constructs the default boto_client and bloop_client if it doesn't already have one

---

For an immediate workaround, you can use the following (it's not pretty, and it will try to make a default boto3 client first, so you'll need valid credentials there)

```python
import boto3
import bloop
from bloop.client import Client

# Talk to DynamoDB Local
boto_client = boto3.client("dynamodb", endpoint_url="http://localhost:8000")

# bloop.client.Client creates a boto3 client on __init__, unfortunately.
bloop_client = Client()
# Replace the default client with the local client above
bloop_client.client = boto_client

engine = bloop.Engine()
engine.client = bloop_client
```

**Comment by numberoverzero** [comment]
Alright, I think that should do it.  I want to mull it over a bit, and get some examples in the docs first.

For now feel free to check out the `client-refactor` branch and you should be able to use the first sample:

```python
import boto3
import bloop

# Talk to DynamoDB Local
boto_client = ddb = boto3.client("dynamodb", endpoint_url="http://localhost:8000")
bloop_client = bloop.Client(boto_client=boto_client)
engine = bloop.Engine(client=bloop_client)
```

**Comment by jasonlockporcelane** [comment]
Awesome, thank you! Will take this for a test drive.

**Comment by numberoverzero** [comment]
Sorry, didn't mean to close this without your feedback.  I'm uploading 0.9.11 to pypi now, can you update (`pip install -U bloop`) and give it a try?

Thanks!

**Comment by numberoverzero** [comment]
The DynamoDB Local sample at [docs/patterns](http://bloop.readthedocs.io/en/latest/user/patterns.html) is effectively identical to the proposed code at the top of this thread.

**Comment by jasonlockporcelane** [comment]
Fantastic. I'll update and give it try now.

**Comment by jasonlockporcelane** [comment]
Thank you for quick turn around on this! For what is worth, I like the implementation. Had some thoughts while testing out 0.9.11. Some might be best suited for another issue, but just wanted to share:  

1. What do you think about updating the [patterns doc](https://github.com/numberoverzero/bloop/blob/master/docs/user/patterns.rst) to include  the `region_name` kwarg when creating a new boto3 client? Without it, botocore raises a `NoRegionError` exception. 

2. The version for [DynamoDB local](https://hub.docker.com/r/deangiberson/aws-dynamodb-local/) i'm using, includes the [LastIncreaseDateTime](http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_ProvisionedThroughputDescription.html) and [LastDecreaseDateTime](http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_ProvisionedThroughputDescription.html) keys in the ProvisionedThroughput dict, causing a `TableMismatch` exception to be raised when the table is validated [here](https://github.com/numberoverzero/bloop/blob/2316164268e1c702dcbb657585277c25c5626769/bloop/client.py#L378). Maybe those keys can be added to [_table_for_model](https://github.com/numberoverzero/bloop/blob/2316164268e1c702dcbb657585277c25c5626769/bloop/client.py#L474) or added to the [junk_index_fields](https://github.com/numberoverzero/bloop/blob/2316164268e1c702dcbb657585277c25c5626769/bloop/client.py#L306) black list in describe_table? Heres the full exception message: 
```python
"Existing table for model <class 'frontend.views.user_patient.create_patient.<locals>.UserPatient'> does not match expected", 
{
  'TableName': 'PatientUser', 
  'KeySchema': [{'KeyType': 'HASH', 'AttributeName': 'id'}], 
  'ProvisionedThroughput': {'ReadCapacityUnits': 5, 'WriteCapacityUnits': 10}, 
  'AttributeDefinitions': [{'AttributeName': 'id', 'AttributeType': 'S'}]
}, 
{
  'TableName': 'PatientUser', 
  'KeySchema': [{'KeyType': 'HASH', 'AttributeName': 'id'}], 
  'ProvisionedThroughput': {
    'ReadCapacityUnits': 1, 
    'WriteCapacityUnits': 1, 
    'LastIncreaseDateTime': datetime.datetime(1970, 1, 1, 0, 0, tzinfo=tzlocal()), 
    'LastDecreaseDateTime': datetime.datetime(1970, 1, 1, 0, 0, tzinfo=tzlocal())}, 
    'AttributeDefinitions': [{'AttributeName': 'id', 'AttributeType': 'S'}]
}
```
  
It's most likely i'm doing something completely wrong, so here is a [gist](https://gist.github.com/jasonlockporcelane/756d1d424f3286958cf5027ecaf92372) of the code i'm currently testing. See any glaring issues?

Happy to help with PRs. Thanks again :)

**Comment by numberoverzero** [comment]
The gist looks fine; I'd recommend setting the region in your aws config so boto3 picks it up automatically.

1. I'll add a region to the docs; you can also fix that by specifying a default region in your aws config.  Boto3 refuses to pick a default region for you, so if there's not one in your config's profile or provided in kwargs, it raises.

    ```
    [profile gaas-integ]
    region = us-west-2
    aws_access_key_id = ...
    aws_secret_access_key = ...
    ```

2. Entirely agreed, I'm opening another issue.  I was pretty sure I got rid of those in a previous pass, but I guess they're not returned by DynamoDB, since tables are verifying correctly for me.

Thanks for the feedback; your descriptions are thorough enough to make fixes easy :guitar:

**Comment by numberoverzero** [comment]
Just released `0.9.12`, which handles extra fields in the response from DynamoDB.  [Here's the test](https://github.com/numberoverzero/bloop/blob/f4dbefd3743b97d15bc46f6b90c00b261911e274/tests/test_tables.py#L112) that ensures extra fields like `"LastDecreaseDateTime"` and `"LastIncreaseDateTime"` are dropped before comparing against the expected table.

If you don't mind, please give `0.9.12` a try and re-open #39 if you're still getting `TableMismatch`.

**Comment by jasonlockporcelane** [comment]
0.9.12 is working like a charm! I really appreciate you taking the time for this. Thanks again.
